File: apps/api/momento_api/processor/metadata.py
import subprocess
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional

from PIL import Image
from PIL.ExifTags import GPSTAGS, TAGS

ExifData = dict[str, object]


def _string_or_none(value: object | None) -> Optional[str]:
    if isinstance(value, str):
        return value
    return None


def _int_or_none(value: object | None) -> Optional[int]:
    if isinstance(value, int):
        return value
    if isinstance(value, str) and value.isdigit():
        return int(value)
    return None


def _float_or_none(value: object | None) -> Optional[float]:
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    if isinstance(value, tuple) and len(value) == 2:
        numerator_value, denominator_value = value
        try:
            return float(numerator_value) / float(denominator_value)
        except (ZeroDivisionError, TypeError, ValueError):
            return None
    return None


def _format_exposure_time(value: object | None) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, str):
        return value
    return None


def _extract_keywords(exif: dict[str, object]) -> Optional[str]:
    keyword_value = exif.get("XPKeywords") or exif.get("Keywords")
    if keyword_value is None:
        return None
    if isinstance(keyword_value, bytes):
        try:
            return keyword_value.decode("utf-16le").strip("\x00")
        except UnicodeDecodeError:
            return None
    if isinstance(keyword_value, str):
        return keyword_value
    return None


def _coerce_exif_mapping(exif_data: Mapping[int, object]) -> ExifData:
    return {str(TAGS.get(k, k)): v for k, v in exif_data.items()}


@dataclass
class MediaMetadata:
    width: Optional[int] = None
    height: Optional[int] = None
    date_taken: Optional[datetime] = None
    gps_latitude: Optional[float] = None
    gps_longitude: Optional[float] = None
    gps_altitude: Optional[float] = None
    camera_make: Optional[str] = None
    camera_model: Optional[str] = None
    iso: Optional[int] = None
    exposure_time: Optional[str] = None
    f_number: Optional[float] = None
    focal_length: Optional[float] = None
    keywords: Optional[str] = None
    duration_seconds: Optional[float] = None
    mime_type: Optional[str] = None
    location_state: Optional[str] = None
    location_country: Optional[str] = None


def _convert_to_degrees(value) -> float:
    d, m, s = value
    return float(d) + float(m) / 60.0 + float(s) / 3600.0


def _parse_exif_datetime(dt_str: str) -> Optional[datetime]:
    formats = ["%Y:%m:%d %H:%M:%S", "%Y-%m-%d %H:%M:%S", "%Y:%m:%d", "%Y-%m-%d"]
    for fmt in formats:
        try:
            return datetime.strptime(dt_str.strip(), fmt)
        except ValueError:
            continue
    return None


def _fallback_to_mtime(file_path: Path) -> Optional[datetime]:
    try:
        return datetime.fromtimestamp(file_path.stat().st_mtime)
    except OSError:
        return None


def _extract_exif_date(exif: dict[str, object]) -> Optional[datetime]:
    date_fields = ["DateTimeOriginal", "DateTimeDigitized", "DateTime"]
    for field in date_fields:
        if field in exif:
            parsed = _parse_exif_datetime(str(exif[field]))
            if parsed:
                return parsed
    return None


def extract_image_metadata(file_path: Path) -> MediaMetadata:
    metadata = MediaMetadata()

    try:
        image = Image.open(file_path)
    except OSError:
        metadata.date_taken = _fallback_to_mtime(file_path)
        return metadata

    with image:
        metadata.width = image.width
        metadata.height = image.height
        img_format = image.format or "JPEG"
        metadata.mime_type = Image.MIME.get(img_format, "image/jpeg")

        exif_data = image.getexif()
        if not exif_data:
            metadata.date_taken = _fallback_to_mtime(file_path)
            return metadata

        exif = _coerce_exif_mapping(exif_data)

        metadata.date_taken = _extract_exif_date(exif) or _fallback_to_mtime(file_path)

        metadata.camera_make = _string_or_none(exif.get("Make"))
        metadata.camera_model = _string_or_none(exif.get("Model"))
        metadata.iso = _int_or_none(exif.get("ISOSpeedRatings"))
        metadata.exposure_time = _format_exposure_time(exif.get("ExposureTime"))
        metadata.f_number = _float_or_none(exif.get("FNumber"))
        metadata.focal_length = _float_or_none(exif.get("FocalLength"))
        metadata.keywords = _extract_keywords(exif)

        gps_data = exif.get("GPSInfo")
        if not isinstance(gps_data, dict):
            return metadata

        gps_info = {str(GPSTAGS.get(k, k)): v for k, v in gps_data.items()}

        if "GPSLatitude" in gps_info and "GPSLatitudeRef" in gps_info:
            lat = _convert_to_degrees(gps_info["GPSLatitude"])
            if gps_info["GPSLatitudeRef"] == "S":
                lat = -lat
            metadata.gps_latitude = lat

        if "GPSLongitude" in gps_info and "GPSLongitudeRef" in gps_info:
            lon = _convert_to_degrees(gps_info["GPSLongitude"])
            if gps_info["GPSLongitudeRef"] == "W":
                lon = -lon
            metadata.gps_longitude = lon

        if "GPSAltitude" in gps_info:
            metadata.gps_altitude = _float_or_none(gps_info["GPSAltitude"])

    return metadata


def extract_video_metadata(file_path: Path) -> MediaMetadata:
    metadata = MediaMetadata()

    try:
        result = subprocess.run(
            ["ffprobe", "-v", "quiet", "-print_format", "json", "-show_format", "-show_streams", str(file_path)],
            capture_output=True,
            text=True,
            timeout=30,
        )
    except (OSError, ValueError):
        metadata.date_taken = _fallback_to_mtime(file_path)
        metadata.duration_seconds = 0.0
        return metadata

    if result.returncode != 0:
        metadata.date_taken = _fallback_to_mtime(file_path)
        metadata.duration_seconds = 0.0
        return metadata

    ffprobe_data = _parse_ffprobe_output(result.stdout)
    if ffprobe_data is None:
        metadata.date_taken = _fallback_to_mtime(file_path)
        return metadata

    _apply_video_stream_metadata(metadata, ffprobe_data)
    _apply_video_format_metadata(metadata, ffprobe_data, file_path)
    return metadata


def _parse_ffprobe_output(output: str) -> Optional[dict[str, object]]:
    import json

    try:
        parsed = json.loads(output)
    except json.JSONDecodeError:
        return None

    if not isinstance(parsed, dict):
        return None
    return parsed


def _apply_video_stream_metadata(metadata: MediaMetadata, data: dict[str, object]) -> None:
    stream_entries = data.get("streams")
    if not isinstance(stream_entries, list):
        return

    for entry in stream_entries:
        if not isinstance(entry, dict):
            continue
        if entry.get("codec_type") != "video":
            continue
        metadata.width = entry.get("width")
        metadata.height = entry.get("height")
        return


def _apply_video_format_metadata(metadata: MediaMetadata, data: dict[str, object], file_path: Path) -> None:
    format_data = data.get("format")
    if not isinstance(format_data, dict):
        metadata.date_taken = _fallback_to_mtime(file_path)
        return

    duration_value = format_data.get("duration")
    if duration_value is None:
        # Try to get duration from video stream if format duration is missing
        stream_entries = data.get("streams")
        if isinstance(stream_entries, list):
            for entry in stream_entries:
                if isinstance(entry, dict) and entry.get("codec_type") == "video" and "duration" in entry:
                    duration_value = entry.get("duration")
                    break

    if duration_value is not None:
        try:
            metadata.duration_seconds = float(duration_value)
        except (TypeError, ValueError):
            metadata.duration_seconds = 0.0
    else:
        metadata.duration_seconds = 0.0

    tag_data = format_data.get("tags")
    if not isinstance(tag_data, dict):
        tag_data = {}

    creation_time_value = tag_data.get("creation_time") or tag_data.get("com.apple.quicktime.creationdate")
    if isinstance(creation_time_value, str):
        try:
            metadata.date_taken = datetime.fromisoformat(creation_time_value.replace("Z", "+00:00"))
        except ValueError:
            metadata.date_taken = None

    if metadata.date_taken is None:
        metadata.date_taken = _fallback_to_mtime(file_path)

    location_value = tag_data.get("location") or tag_data.get("com.apple.quicktime.location.ISO6709")
    if isinstance(location_value, str):
        metadata.gps_latitude, metadata.gps_longitude = _parse_iso6709_location(location_value)

    ext = file_path.suffix.lower()
    mime_map = {
        ".mp4": "video/mp4",
        ".mov": "video/quicktime",
        ".avi": "video/x-msvideo",
        ".mkv": "video/x-matroska",
        ".webm": "video/webm",
        ".m4v": "video/x-m4v",
    }
    metadata.mime_type = mime_map.get(ext, "video/mp4")


def _parse_iso6709_location(location: str) -> tuple[Optional[float], Optional[float]]:
    location_value = location.rstrip("/")
    if "+" not in location_value[1:] and "-" not in location_value[1:]:
        return None, None

    lat_str = ""
    lon_str = ""
    for index, character in enumerate(location_value[1:], 1):
        if character in "+-":
            lat_str = location_value[:index]
            lon_str = location_value[index:]
            break

    if not lat_str or not lon_str:
        return None, None

    try:
        lat = float(lat_str)
        lon_end = lon_str.find("+", 1) if "+" in lon_str[1:] else lon_str.find("-", 1)
        if lon_end == -1:
            lon = float(lon_str)
        else:
            lon = float(lon_str[:lon_end])
        return lat, lon
    except ValueError:
        return None, None

File: apps/api/src/processor/metadata.rs
use chrono::{DateTime, NaiveDateTime, Utc};
use image::ImageReader;
use serde::Deserialize;
use std::path::Path;
use std::process::Command;

#[derive(Debug, Default, Clone)]
pub struct MediaMetadata {
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub date_taken: Option<DateTime<Utc>>,
    pub gps_latitude: Option<f64>,
    pub gps_longitude: Option<f64>,
    pub gps_altitude: Option<f64>,
    pub camera_make: Option<String>,
    pub camera_model: Option<String>,
    pub iso: Option<i32>,
    pub exposure_time: Option<String>,
    pub f_number: Option<f64>,
    pub focal_length: Option<f64>,
    pub keywords: Option<String>,
    pub duration_seconds: Option<f64>,
    pub mime_type: Option<String>,
    pub location_state: Option<String>,
    pub location_country: Option<String>,
}

fn fallback_to_mtime(file_path: &Path) -> Option<DateTime<Utc>> {
    file_path
        .metadata()
        .ok()
        .and_then(|m| m.modified().ok())
        .map(|t| DateTime::<Utc>::from(t))
}

pub fn extract_image_metadata(file_path: &Path) -> MediaMetadata {
    let mut metadata = MediaMetadata::default();

    // Try to read basic image info
    match ImageReader::open(file_path) {
        Ok(reader) => {
            if let Ok(img) = reader.decode() {
                metadata.width = Some(img.width() as i32);
                metadata.height = Some(img.height() as i32);
            }
        }
        Err(_) => {
            metadata.date_taken = fallback_to_mtime(file_path);
            return metadata;
        }
    }

    // Determine MIME type from extension
    let ext = file_path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();

    metadata.mime_type = Some(match ext.as_str() {
        "jpg" | "jpeg" => "image/jpeg",
        "png" => "image/png",
        "gif" => "image/gif",
        "webp" => "image/webp",
        "heic" | "heif" => "image/heic",
        "tiff" => "image/tiff",
        "bmp" => "image/bmp",
        _ => "image/jpeg",
    }.to_string());

    // Try to read EXIF data using exiftool (more reliable than pure Rust)
    if let Ok(output) = Command::new("exiftool")
        .args(["-json", "-n", file_path.to_str().unwrap_or("")])
        .output()
    {
        if output.status.success() {
            if let Ok(json_str) = String::from_utf8(output.stdout) {
                if let Ok(exif_data) = serde_json::from_str::<Vec<ExifToolOutput>>(&json_str) {
                    if let Some(data) = exif_data.into_iter().next() {
                        apply_exif_data(&mut metadata, data);
                    }
                }
            }
        }
    }

    if metadata.date_taken.is_none() {
        metadata.date_taken = fallback_to_mtime(file_path);
    }

    metadata
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct ExifToolOutput {
    #[serde(alias = "DateTimeOriginal")]
    date_time_original: Option<String>,
    #[serde(alias = "CreateDate")]
    create_date: Option<String>,
    #[serde(alias = "GPSLatitude")]
    gps_latitude: Option<f64>,
    #[serde(alias = "GPSLongitude")]
    gps_longitude: Option<f64>,
    #[serde(alias = "GPSAltitude")]
    gps_altitude: Option<f64>,
    make: Option<String>,
    model: Option<String>,
    #[serde(alias = "ISO")]
    iso: Option<i32>,
    #[serde(alias = "ExposureTime")]
    exposure_time: Option<f64>,
    #[serde(alias = "FNumber")]
    f_number: Option<f64>,
    #[serde(alias = "FocalLength")]
    focal_length: Option<f64>,
    #[serde(alias = "Keywords")]
    keywords: Option<serde_json::Value>,
    #[serde(alias = "ImageWidth")]
    image_width: Option<i32>,
    #[serde(alias = "ImageHeight")]
    image_height: Option<i32>,
}

fn apply_exif_data(metadata: &mut MediaMetadata, data: ExifToolOutput) {
    // Date
    let date_str = data.date_time_original.or(data.create_date);
    if let Some(date_str) = date_str {
        metadata.date_taken = parse_exif_datetime(&date_str);
    }

    // GPS
    metadata.gps_latitude = data.gps_latitude;
    metadata.gps_longitude = data.gps_longitude;
    metadata.gps_altitude = data.gps_altitude;

    // Camera info
    metadata.camera_make = data.make;
    metadata.camera_model = data.model;
    metadata.iso = data.iso;
    metadata.f_number = data.f_number;
    metadata.focal_length = data.focal_length;

    // Exposure time as string
    if let Some(exp) = data.exposure_time {
        if exp > 0.0 && exp < 1.0 {
            metadata.exposure_time = Some(format!("1/{}", (1.0 / exp).round() as i32));
        } else {
            metadata.exposure_time = Some(format!("{}", exp));
        }
    }

    // Keywords
    if let Some(kw) = data.keywords {
        metadata.keywords = match kw {
            serde_json::Value::String(s) => Some(s),
            serde_json::Value::Array(arr) => {
                let strs: Vec<String> = arr
                    .iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect();
                if strs.is_empty() {
                    None
                } else {
                    Some(strs.join(","))
                }
            }
            _ => None,
        };
    }

    // Dimensions from EXIF if not already set
    if metadata.width.is_none() {
        metadata.width = data.image_width;
    }
    if metadata.height.is_none() {
        metadata.height = data.image_height;
    }
}

fn parse_exif_datetime(dt_str: &str) -> Option<DateTime<Utc>> {
    // Try common formats
    let formats = [
        "%Y:%m:%d %H:%M:%S",
        "%Y-%m-%d %H:%M:%S",
        "%Y:%m:%d",
        "%Y-%m-%d",
    ];

    let clean_str = dt_str.trim();
    for fmt in &formats {
        if let Ok(naive) = NaiveDateTime::parse_from_str(clean_str, fmt) {
            return Some(DateTime::from_naive_utc_and_offset(naive, Utc));
        }
    }

    None
}

pub fn extract_video_metadata(file_path: &Path) -> MediaMetadata {
    let mut metadata = MediaMetadata::default();

    // Run ffprobe
    let output = Command::new("ffprobe")
        .args([
            "-v", "quiet",
            "-print_format", "json",
            "-show_format",
            "-show_streams",
            file_path.to_str().unwrap_or(""),
        ])
        .output();

    let output = match output {
        Ok(o) if o.status.success() => o,
        _ => {
            metadata.date_taken = fallback_to_mtime(file_path);
            metadata.duration_seconds = Some(0.0);
            return metadata;
        }
    };

    let json_str = match String::from_utf8(output.stdout) {
        Ok(s) => s,
        Err(_) => {
            metadata.date_taken = fallback_to_mtime(file_path);
            return metadata;
        }
    };

    let ffprobe_data: FfprobeOutput = match serde_json::from_str(&json_str) {
        Ok(d) => d,
        Err(_) => {
            metadata.date_taken = fallback_to_mtime(file_path);
            return metadata;
        }
    };

    // Extract video stream info
    if let Some(streams) = ffprobe_data.streams {
        for stream in streams {
            if stream.codec_type.as_deref() == Some("video") {
                metadata.width = stream.width;
                metadata.height = stream.height;
                break;
            }
        }
    }

    // Extract format info
    if let Some(format) = ffprobe_data.format {
        // Duration
        if let Some(duration) = format.duration {
            metadata.duration_seconds = duration.parse().ok();
        }

        // Tags
        if let Some(tags) = format.tags {
            // Creation time
            let creation_time = tags.creation_time.or(tags.com_apple_quicktime_creationdate);
            if let Some(ct) = creation_time {
                let clean_ct = ct.replace("Z", "+00:00");
                if let Ok(dt) = DateTime::parse_from_rfc3339(&clean_ct) {
                    metadata.date_taken = Some(dt.with_timezone(&Utc));
                }
            }

            // Location
            let location = tags.location.or(tags.com_apple_quicktime_location_iso6709);
            if let Some(loc) = location {
                if let Some((lat, lon)) = parse_iso6709_location(&loc) {
                    metadata.gps_latitude = Some(lat);
                    metadata.gps_longitude = Some(lon);
                }
            }
        }
    }

    // Fallback date
    if metadata.date_taken.is_none() {
        metadata.date_taken = fallback_to_mtime(file_path);
    }

    // MIME type from extension
    let ext = file_path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();

    metadata.mime_type = Some(match ext.as_str() {
        "mp4" => "video/mp4",
        "mov" => "video/quicktime",
        "avi" => "video/x-msvideo",
        "mkv" => "video/x-matroska",
        "webm" => "video/webm",
        "m4v" => "video/x-m4v",
        _ => "video/mp4",
    }.to_string());

    metadata
}

#[derive(Debug, Deserialize)]
struct FfprobeOutput {
    streams: Option<Vec<FfprobeStream>>,
    format: Option<FfprobeFormat>,
}

#[derive(Debug, Deserialize)]
struct FfprobeStream {
    codec_type: Option<String>,
    width: Option<i32>,
    height: Option<i32>,
}

#[derive(Debug, Deserialize)]
struct FfprobeFormat {
    duration: Option<String>,
    tags: Option<FfprobeTags>,
}

#[derive(Debug, Deserialize)]
struct FfprobeTags {
    creation_time: Option<String>,
    #[serde(rename = "com.apple.quicktime.creationdate")]
    com_apple_quicktime_creationdate: Option<String>,
    location: Option<String>,
    #[serde(rename = "com.apple.quicktime.location.ISO6709")]
    com_apple_quicktime_location_iso6709: Option<String>,
}

fn parse_iso6709_location(location: &str) -> Option<(f64, f64)> {
    let location = location.trim_end_matches('/');
    if location.len() < 2 {
        return None;
    }

    // Find second +/- after position 1
    let chars: Vec<char> = location.chars().collect();
    let mut split_idx = 0;

    for (i, &c) in chars.iter().enumerate().skip(1) {
        if c == '+' || c == '-' {
            split_idx = i;
            break;
        }
    }

    if split_idx == 0 {
        return None;
    }

    let lat_str: String = chars[..split_idx].iter().collect();
    let mut lon_str: String = chars[split_idx..].iter().collect();

    // Handle altitude suffix
    if let Some(pos) = lon_str[1..].find(|c: char| c == '+' || c == '-') {
        lon_str = lon_str[..pos + 1].to_string();
    }

    let lat: f64 = lat_str.parse().ok()?;
    let lon: f64 = lon_str.parse().ok()?;

    Some((lat, lon))
}
